MessageProcessorModule/src/main/java/com/lotus/iot/mpm/handler/MessageHandler.java

```diff
@@ -0,0 +1,55 @@
+package com.lotus.iot.mpm.handler;
+
+import com.lotus.iot.mpm.api.objectmodel.response.GeneralResponse;
+import com.lotus.iot.mpm.api.objectmodel.response.GeneralResponseCode;
+import com.lotus.iot.mpm.datamodel.service.AssetStatusService;
+import com.lotus.iot.mpm.eventhandler.EventHandlerFacade;
+import com.lotus.iot.mpm.exception.OldMessageException;
+import com.lotus.iot.mpm.objectmodel.AssetStatusMessage;
+import com.lotus.iot.mpm.objectmodel.type.StatusType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import java.util.Date;
+
+/**
+ * @author Niloufar Mazloumpour
+ * @mail niloufar@mazloumpour.net
+ * @since 2018-05-22
+ */
+
+@Component
+public class MessageHandler {
+    private static final Logger LOGGER = LoggerFactory.getLogger(MessageHandler.class);
+
+    private AssetStatusService assetStatusService;
+    private EventHandlerFacade eventHandlerFacade;
+
+    @Autowired
+    public MessageHandler(AssetStatusService assetStatusService, EventHandlerFacade eventHandlerFacade) {
+        this.assetStatusService = assetStatusService;
+        this.eventHandlerFacade = eventHandlerFacade;
+    }
+
+    public GeneralResponse assetStatus(AssetStatusMessage assetStatusMessage) {
+        try {
+            //store asset status
+            StatusType oldStatusType = assetStatusService.updateStatus(assetStatusMessage);
```

Comment:
There are two collaborators in this method one is `AssetStatusService` which contains asset statuses and is essentially a _"state object"_. The second collaborator here is `EventHandlerFacade` which is used to notify observers/listeners about state changes. Observers use `EventHandlerFacade` to subscribe to events hence it is a _"Subject"_ that observers listening to. Since all the state is stored in `AssetStatusService` and the `EventHandlerFacade` is not aware of those state changes this `MessageHandler` class is used to bridge the gap between the two collaborators.

I see a discrepancy here. Usually, the _"Subject"_ is something that is aware of the state and can produce those events about the state change. I think in this case the implementation would be simpler if the observers would subscribe to events on `AssetStatusService` which would be able to produce those status change events. Then we probably wouldn't need this `MessageHandler` class as well as the `EventHandlerFacade`.

Reply:
I totally agree with you. Thank you.